File: src/ks0223-web-mac/backend/Services/RuntimeModes.cs
namespace Ks0223.Web.Backend.Services;

public static class RuntimeModes
{
    public const string RealRobot = "real-robot";
    public const string UnitySim = "unity-sim";

    public static string Normalize(string? mode)
    {
        if (string.IsNullOrWhiteSpace(mode))
        {
            return RealRobot;
        }

        return mode.Trim().ToLowerInvariant() switch
        {
            "real" or "real-robot" or "pi" or "robot" => RealRobot,
            "unity" or "unity-sim" or "sim" or "simulator" => UnitySim,
            _ => mode.Trim().ToLowerInvariant(),
        };
    }
}

File: src/ks0223-web-mac/frontend/src/runtime-modes.ts
/**
 * Runtime mode the operator UI is talking to.
 *
 *   - `real-robot`: physical KS0223 reachable via the backend Pi-bridge.
 *   - `unity-sim` : Unity simulator instance hosting the runtime HTTP API.
 *
 * Kept in its own module so cross-cutting helpers (storage keys, defaults,
 * connection card) can refer to it without depending on `App.tsx`.
 */
export type RuntimeMode = 'real-robot' | 'unity-sim'

export const RUNTIME_MODES: readonly RuntimeMode[] = ['real-robot', 'unity-sim'] as const

export function isRuntimeMode(value: unknown): value is RuntimeMode {
  return value === 'real-robot' || value === 'unity-sim'
}

export function normalizeRuntimeMode(
  value: unknown,
  fallback: RuntimeMode = 'real-robot',
): RuntimeMode {
  return isRuntimeMode(value) ? value : fallback
}
